### DISCUSSION TITLE
Can I use PI4J to create and use virtual serial ports on Raspberry Pi?

### DISCUSSION THREAD

**Discussion by Conmi-WhiteJoker** [Pi4J/pi4j#97] [Q&A] [answered]
I hope to use Raspberry Pi as the test machine, but the number of serial ports is only 1, adding STM32 will increase the amount of code, I would like to ask if you can set some pins as serial ports through PI4J?
Thanks.

**Comment by Conmi-WhiteJoker** [comment]
Can someone give me a hand?

**Comment by FDelporte** [comment]
From first quick look it seems the underlying PiGpio expects a device name, not a pin number.

https://github.com/Pi4J/pi4j-v2/blob/main/libraries/pi4j-library-pigpio/src/main/java/com/pi4j/library/pigpio/PiGpio_Serial.java#L54
`serOpen(CharSequence device, int baud)`

So a solution could be to use USB-connected serial adapters, which would be listed as `/dev/ttyXYZ`. Could this be an approach for you?

**Reply by Conmi-WhiteJoker** [reply]
I probably mean more like a soft serial port like the Arduino.
My idea is more like having two of the idle pins set up as a serial port, rather than operating on an existing serial port.
`SoftwareSerial myArduinoSerial= SoftwareSerial(rxPin, txPin)`

**Comment by savageautomate** [comment, marked as answer]
When you say "_virtual_" serial ports, do you mean "bit-banged" serial ports?  (The ability to emulate serial communication, signaling and timing via software over arbitrary GPIO pins).   The underlying PiGpio library does support bit-banded serial ports using custom GPIO pins to some degree.  I don't think this is exposed via Pi4J however at this time.   

If you are just looking for expandability, like @FDelporte mentioned, you can easily add USB to serial port adapters to expand the serial capability of the RaspberryPi. 

If by "_virtual_" serial port you mean a serial port exposed over the network, then I would suggest using the `ser2net` utility:
https://linux.die.net/man/8/ser2net#:~:text=The%20ser2net%20daemon%20allows%20telnet,and%20open%20the%20serial%20port

**Reply by Conmi-WhiteJoker** [reply]
I probably mean more like a soft serial port like the Arduino.I wonder if there is a similar implementation of Raspberry Pi?
`SoftwareSerial myArduinoSerial= SoftwareSerial(rxPin, txPin)`

**Comment by savageautomate** [comment]
I realize this is an old thread but I'm just responding for any future readers.

---

The RaspberryPi 3 has a couple of UARTS where you can access at least two serial ports (_with some notable caveats_) 
SEE: https://docs.bitscope.com/pi-serial/

The RPi4 has an additional 4 UARTS for serial communication.

These additional ports require special configuration but when properly configured, Pi4J can allow serial communication via the ports.  

At this time Pi4J does not directly support any "virtual" / "software serial" / "bit-banged" serial ports via its APIs.

The PiGPIO library does provide some limited bit-banged (software) serial communication and you couple probably use the underlying Pi4J PiGPIO library directly to try and use these functions from Java.  See:
https://github.com/Pi4J/pi4j-v2/tree/develop/libraries/pi4j-library-pigpio/src/main/java/com/pi4j/library/pigpio
Use at your own risk as software implemented serial ports can be buggy.

Thanks, Robert
